python/cudf/cudf/core/index.py

```diff
@@ -178,6 +178,36 @@ def serialize(self):
     def __contains__(self, item):
         return item in self._values
 
+    @annotate("INDEX_EQUALS", color="green", domain="cudf_python")
+    def equals(self, other, **kwargs):
+        """
+        Determine if two Index objects contain the same elements.
+
+        Returns
+        -------
+        out: bool
+            True if “other” is an Index and it has the same elements
+            as calling index; False otherwise.
+        """
+        if not isinstance(other, Index):
+            return False
+
+        check_types = False
+
+        self_is_categorical = isinstance(self, CategoricalIndex)
+        other_is_categorical = isinstance(other, CategoricalIndex)
+        if self_is_categorical and not other_is_categorical:
+            other = other.astype(self.dtype)
+            check_types = True
+        elif other_is_categorical and not self_is_categorical:
+            self = self.astype(other.dtype)
+            check_types = True
```

Comment:
I'm pretty sure if the dtypes differ Pandas just always returns False:
```
import pandas as pd

pdi1 = pd.CategoricalIndex([0, 1, 2, 3, 4], categories=[1, 2, 3, 4, 5])
pdi2 = pd.Int64Index([1, 2, 3, 4, 5])
pdi3 = pd.CategoricalIndex([1, 2, 3, 4, 5], categories=[0, 1, 2, 3, 4, 5])

print(pdi1.equals(pdi2))
# False
print(pdi1.equals(pdi3))
# False
```

Reply:
Right, only when categoricals are involved it looks like:

```
In [10]: a                                  
Out[10]: Float64Index([1.0, 2.0, 3.0, 4.0, 5.0], dtype='float64')

In [11]: b                                  
Out[11]: Int64Index([1, 2, 3, 4, 5], dtype='int64')

In [12]: a.equals(b)                        
Out[12]: True
```

Reply:
Yea, if I'm not mistaken your code above typecasts when either `a` or `b` are a `CategoricalIndex`

Reply:
Hmm I think maybe your example had a typo. Did you mean to compare the following?

```python
import pandas as pd                                                                                                                                                                                  
pdi1 = pd.CategoricalIndex([1, 2, 3, 4, 5], categories=[1, 2, 3, 4, 5])
pdi2 = pd.Int64Index([1, 2, 3, 4, 5])
pdi3 = pd.CategoricalIndex([1, 2, 3, 4, 5], categories=[0, 1, 2, 3, 4, 5])                                                                                                                                                                                 

print(pdi1.equals(pdi2))
# True
print(pdi1.equals(pdi3)
# False
```

Reply:
Yup I had a typo, good catch, sorry!

Reply:
Resolving this conversation.